File: leet_0552_attendance.cc
class Solution {
public:
    int checkRecord(int N)
    {
        vector<vector<vector<int>>> cache(N, vector<vector<int>>(2, vector<int>(3, 0)));
        vector<vector<vector<bool>>> hasCache(N, vector<vector<bool>>(2, vector<bool>(3, false)));

        int MOD = int(1e9 + 7);

        function<int(int, int, int)> DFS = [&](int days, int A, int L)
        {
            // base and caching
            if (days == N)
            return 1;
            if (hasCache[days][A][L]) return cache[days][A][L];

            int res = 0;

            // present - nothing happens
            res += DFS(days + 1, A, 0);
            res %= MOD;

            // absent - add to A where max(A) == 1
            if (A == 0) {
                res += DFS(days + 1, A + 1, 0);
                res %= MOD;
            }

            // late - add to consecutive L where max(L) == 2
            if (L < 2) {
                res += DFS(days + 1, A, L + 1);
                res %= MOD;
            }

            cache[days][A][L] = res;
            hasCache[days][A][L] = true;

            return res;
        };

        return DFS(0, 0, 0);
    }
};

File: leet_0552_attendance.py
class Solution:
    def checkRecord(self, n: int) -> int:
        MOD = 10 ** 9 + 7

        days, A, L = 0, 0, 0
        cache = [[ [0] * 3 for _ in range(2) ] for _ in range(n)]
        has_cache = [[[False] * 3 for _ in range(2)] for _ in range(n)]

        def DFS(days, A, L):
            # base and caching
            if days == n:
                return 1
            if has_cache[days][A][L]:
                return cache[days][A][L]
            res = 0
            # present - nothing happens
            res += DFS(days + 1, A, 0)
            # absent - add to A where max(A) == 1
            if A == 0:
                res += DFS(days + 1, A + 1, 0)
            # late - add to consecutive L where max(L) == 2
            if L < 2:
                res += DFS(days + 1, A, L + 1)
            res %= MOD
            cache[days][A][L] = res
            has_cache[days][A][L] = True
            return res
        return DFS(days, A, L)


